go/engine/wot_attest.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2020 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"strings"
+
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	jsonw "github.com/keybase/go-jsonw"
+)
+
+type WotAttestArg struct {
+	Attestee     keybase1.UserVersion
+	Confidence   keybase1.Confidence
+	Attestations []string
+}
+
+// WotAttest is an engine.
+type WotAttest struct {
+	arg *WotAttestArg
+	libkb.Contextified
+}
+
+// NewWotAttest creates a WotAttest engine.
+func NewWotAttest(g *libkb.GlobalContext, arg *WotAttestArg) *WotAttest {
+	return &WotAttest{
+		arg:          arg,
+		Contextified: libkb.NewContextified(g),
+	}
+}
+
+// Name is the unique engine name.
+func (e *WotAttest) Name() string {
+	return "WotAttest"
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *WotAttest) Prereqs() Prereqs {
+	return Prereqs{Device: true}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *WotAttest) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *WotAttest) SubConsumers() []libkb.UIConsumer {
+	return nil
+}
+
+// Run starts the engine.
+func (e *WotAttest) Run(mctx libkb.MetaContext) error {
+	luArg := libkb.NewLoadUserArgWithMetaContext(mctx).WithUID(e.arg.Attestee.Uid)
+	them, err := libkb.LoadUser(luArg)
+	if err != nil {
+		return err
+	}
+
+	if them.GetCurrentEldestSeqno() != e.arg.Attestee.EldestSeqno {
+		return errors.New("attestee has reset, make sure you still know them")
+	}
+
+	statement := jsonw.NewDictionary()
+	statement.SetKey("user", them.ToWotStatement())
+	statement.SetKey("attestation", libkb.JsonwStringArray(e.arg.Attestations))
+	statement.SetKey("confidence", e.confidence())
+
+	attest := jsonw.NewDictionary()
+	attest.SetKey("obj", statement)
+	randKey, err := libkb.RandBytes(16)
+	if err != nil {
+		return err
+	}
+	hexKey := hex.EncodeToString(randKey)
+	attest.SetKey("key", jsonw.NewString(hexKey))
+
+	marshaled, err := statement.Marshal()
+	if err != nil {
+		return err
+	}
+
+	mac := hmac.New(sha256.New, randKey)
+	mac.Write(marshaled)
+	sum := mac.Sum(nil)
+
+	expansions := jsonw.NewDictionary()
+	expansions.SetKey(hex.EncodeToString(sum), attest)
+
+	signingKey, err := mctx.G().ActiveDevice.SigningKey()
+	if err != nil {
+		return err
+	}
+
+	me, err := libkb.LoadMe(libkb.NewLoadUserArgWithMetaContext(mctx).WithForceReload())
```

Comment:
Do we need a full Me or is it good enough to just load from active device? If not can use FullSelfCacher?

I try to avoid these since they are especially slow for chris, etc.

Reply:
In order to reuse the code in kbsig.go for ProofMetadata, we need a user.  This isn't going to be a frequent operation, but I'll switch to FullSelfCacher.

Reply:
Makes sense. The fullselfcacher likely will not be feasible either then, in which case, it's fine.

Reply:
i had to use WithSelfForcePoll to get it to work.